app/src/main/java/com/example/pokemon/presenter/FavoritePresenterImpl.kt

```diff
@@ -0,0 +1,46 @@
+package com.example.pokemon.presenter
+
+import android.util.Log
+import com.example.pokemon.MyApplication
+import com.example.pokemon.contract.FavoritesContract
+import com.example.pokemon.data.db.AppDatabase
+import com.example.pokemon.data.repository.PokemonRepository
+import com.example.pokemon.model.Pokemon
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.schedulers.Schedulers
+
+class FavoritePresenterImpl(private var favoriteView: FavoritesContract.FavoriteView?) :
+    FavoritesContract.FavoritePresenter {
+
+    private var pokeList: ArrayList<Pokemon> = arrayListOf()
+    private val appDatabase = AppDatabase.getAppDataBase(MyApplication.application.applicationContext)
+    private val pokemonRepository = PokemonRepository.getInstance(appDatabase)
+    private val compositeDisposable = CompositeDisposable()
+
+    override fun getFavoriteList() {
+        compositeDisposable.add(
+            pokemonRepository.getFavoritePokemonsList().subscribeOn(Schedulers.io()).observeOn(
+                AndroidSchedulers.mainThread()
+            ).subscribe({
+                it?.let {
+                    pokeList = ArrayList(it)
+                    favoriteView?.setPokemonAdapter(pokeList)
+                }
+            }, {
+                favoriteView?.showErrorToast()
+            })
+        )
+    }
+
+    override fun setFavorite(pokemon: Pokemon, buttonState: Boolean) = pokemonRepository.setFavoritePokemon(pokemon, buttonState)
+
+    override fun getPokemon(pos: Int) = pokeList[pos]
+
+    override fun onDestroy() {
+        favoriteView = null
+        compositeDisposable.dispose()
+        AppDatabase.destroyDataBase()
```

Comment:
we don't need to close AppDatabase on fav destroy it is only required in main because that is our activity which lies in the memory all the time